### PR TITLE
[Fix](hdfs-writer) Fix hdfs file writer core with `check failed: _ref_cnt == 0` in dtor of `HdfsFileWriter`.

### PR DESCRIPTION
**Pull request by kaka11chen**
## Issue:
```
F20240421 17:14:37.494115 184986 hdfs_util.h:65] Check failed: _ref_cnt == 0 
*** Check failure stack trace: ***
F20240421 17:14:37.505879 185108 hdfs_util.h:65] Check failed: _ref_cnt == 0 
*** Check failure stack trace: ***
    @     0x556f5236d316  google::LogMessageFatal::~LogMessageFatal()
    @     0x556f5236d316  google::LogMessageFatal::~LogMessageFatal()
    @     0x556f2830e200  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f2830e200  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f2830e21e  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f2830e21e  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f507893b0  doris::vectorized::VHivePartitionWriter::~VHivePartitionWriter()
    @     0x556f507893b0  doris::vectorized::VHivePartitionWriter::~VHivePartitionWriter()
    @     0x556f506c005e  std::_Hashtable<>::clear()
    @     0x556f506c005e  std::_Hashtable<>::clear()
    @     0x556f50780f4f  doris::vectorized::VHiveTableWriter::close()
    @     0x556f50780f4f  doris::vectorized::VHiveTableWriter::close()
    @     0x556f5072bc4f  doris::vectorized::AsyncResultWriter::process_block()
    @     0x556f5072bc4f  doris::vectorized::AsyncResultWriter::process_block()
    @     0x556f5072cd01  std::_Function_handler<>::_M_invoke()
    @     0x556f5072cd01  std::_Function_handler<>::_M_invoke()
    @     0x556f2b02b73d  doris::ThreadPool::dispatch_thread()
    @     0x556f2b02b73d  doris::ThreadPool::dispatch_thread()
    @     0x556f2b008d59  doris::Thread::supervise_thread()
    @     0x556f2b008d59  doris::Thread::supervise_thread()
    @     0x7f2c2bfb4609  start_thread
    @     0x7f2c2bfb4609  start_thread
    @     0x7f2c2c261133  clone
    @     0x7f2c2c261133  clone
    @              (nil)  (unknown)
*** Query id: ac4f457c003d4489-b04ac56ef05b12f0 ***
*** is nereids: 1 ***
*** tablet id: 0 ***
*** Aborted at 1713690877 (unix time) try "date -d @1713690877" if you are using GNU date ***
*** Current BE git commitID: e6f4a2f ***
*** SIGABRT unknown detail explain (@0x38a6) received by PID 14502 (TID 184986 OR 0x7f21d0614700) from PID 14502; stack trace: ***
    @              (nil)  (unknown)
F20240421 17:14:37.505879 185108 hdfs_util.h:65] Check failed: _ref_cnt == 0 F20240421 17:14:37.887202 185110 hdfs_util.h:65] Check failed: _ref_cnt == 0 
*** Check failure stack trace: ***
    @     0x556f5236d316  google::LogMessageFatal::~LogMessageFatal()
    @     0x556f2830e200  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f2830e21e  doris::io::HdfsFileWriter::~HdfsFileWriter()
    @     0x556f507893b0  doris::vectorized::VHivePartitionWriter::~VHivePartitionWriter()
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/doris/be/src/common/signal_handler.h:421
```
The root cause is
When it cannot be processed in the cache (such as when the cache is full), we will create a `fs_handler`, and the life cycle of `fs_handler` is managed by caller. We have separated the hdfs writer and can create `fs_handler` separately, so `HdfsFileSystem` and `HdfsFileWriter` may be callers. In `HdfsFileSystem`, the `fs_handler` of `HdfsFileWriter` is shared, so it needs to be changed to `shared_ptr`.

### Solution
Fix hdfs file writer core with `check failed: _ref_cnt == 0` in dtor of `HdfsFileWriter`.
Change `fs_handler` ptr to `shared_ptr` and remove ref count operations.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kaka11chen** [comment]
run buildall

File: be/src/io/hdfs_util.cpp [hunk 1]
```diff
@@ -103,16 +103,16 @@ void HdfsHandlerCache::_clean_oldest() {
...
     {
         std::lock_guard<std::mutex> l(_lock);
         auto it = _cache.find(hash_code);
         if (it != _cache.end()) {
-            HdfsHandler* handle = it->second.get();
+            std::shared_ptr<HdfsHandler> handle = it->second;
             if (!handle->invalid()) {
                 handle->inc_ref();
-                *fs_handle = handle;
+                *fs_handle = std::move(handle);
```
Comment (gavinchou) [inline be/src/io/hdfs_util.cpp:115]: why do we still have to deal with `inc_ref()` even with shared_ptr support? 
can we integrate the process (inc_ref, dec_ref) into  shared_ptr ?

Comment (kaka11chen) [reply be/src/io/hdfs_util.cpp:115]: It is for fs handle cache.

File: be/src/io/fs/hdfs_file_system.h [hunk 2]
```diff
@@ -88,9 +88,7 @@ class HdfsFileSystem final : public RemoteFileSystem {
                    RuntimeProfile* profile, std::string root_path);
     const THdfsParams& _hdfs_params; // Only used in init, so we can use reference here
     std::string _fs_name;
-    // do not use std::shared_ptr or std::unique_ptr
-    // _fs_handle is managed by HdfsFileSystemCache
-    HdfsHandler* _fs_handle = nullptr;
+    std::shared_ptr<HdfsHandler> _fs_handle = nullptr;
```
Comment (platoneko) [inline be/src/io/fs/hdfs_file_system.h:91]:
```suggestion
    std::shared_ptr<HdfsHandler> _fs_handle;
```
可以不用显式写 = nullptr，智能指针默认初始化都是 nullptr

File: be/src/io/fs/hdfs_file_writer.cpp [hunk 3]
```diff
@@ -45,7 +45,7 @@ bvar::Adder<uint64_t> hdfs_file_being_written("hdfs_file_writer_file_being_writt
 
 static constexpr size_t MB = 1024 * 1024;
 
-HdfsFileWriter::HdfsFileWriter(Path path, HdfsHandler* handler, hdfsFile hdfs_file,
+HdfsFileWriter::HdfsFileWriter(Path path, std::shared_ptr<HdfsHandler> handler, hdfsFile hdfs_file,
                                std::string fs_name, const FileWriterOptions* opts)
         : _path(std::move(path)),
           _hdfs_handler(handler),
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/io/fs/hdfs_file_writer.cpp:51]: ditto:
    return std::make_unique<HdfsFileWriter>(std::move(path), std::move(handler), hdfs_file, fs_name, opts);

File: be/src/io/hdfs_util.cpp [hunk 4]
```diff
@@ -129,12 +129,12 @@ Status HdfsHandlerCache::get_connection(const THdfsParams& hdfs_params, const st
             _clean_oldest();
         }
         if (_cache.size() < MAX_CACHE_HANDLE) {
-            std::unique_ptr<HdfsHandler> handle = std::make_unique<HdfsHandler>(hdfs_fs, true);
+            std::shared_ptr<HdfsHandler> handle = std::make_shared<HdfsHandler>(hdfs_fs, true);
```
Comment (platoneko) [inline be/src/io/hdfs_util.cpp:132]:
```suggestion
            auto handle = std::make_shared<HdfsHandler>(hdfs_fs, true);
```

**Review by platoneko — CHANGES_REQUESTED** [review]

**Comment by kaka11chen** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by platoneko — APPROVED** [review]

**Pull request merged**